File: Algorithm/Java/Binary Tree Maximum Path Sum.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private static class ResultType {
        int singlePath;
        int maxPath;
        ResultType(int singlePath, int maxPath) {
            this.singlePath = singlePath;
            this.maxPath = maxPath;
        }
    }
    private ResultType helper(TreeNode root) {
        if (root == null) {
            return new ResultType(0, Integer.MIN_VALUE);
        }
        ResultType left = helper(root.left);
        ResultType right = helper(root.right);
        int singlePath = Math.max(left.singlePath, right.singlePath) + root.val;
        singlePath = Math.max(singlePath, 0);
        int maxPath = Math.max(left.maxPath, right.maxPath);
        maxPath = Math.max(maxPath, left.singlePath + root.val + right.singlePath);
        return new ResultType(singlePath, maxPath);
    }
    public int maxPathSum(TreeNode root) {
        return helper(root).maxPath;
    }
}

File: Algorithm/JavaScript/Binary Tree Maximum Path Sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
    if (!root) {
        return 0;
    }

    const helper = (root) => {
        if (!root) {
            return [0, Number.NEGATIVE_INFINITY];
        }

        const left = helper(root.left);
        const right = helper(root.right);

        const singlePath = root.val + Math.max(left[0], right[0]);
        const maxPath = Math.max(Math.max(left[1], right[1]), left[0] + root.val + right[0]);

        return [Math.max(0, singlePath), maxPath];
    }

    return helper(root)[1];
};